### PR TITLE
Core/725 - Get state_province_id given non numeric country_id or country parameter

### PR DESCRIPTION
**Pull request by jackrabbithanna**
Overview
----------------------------------------
Part Duex for https://lab.civicrm.org/dev/core/issues/725

ping @totten and @eileenmcnaughton 

Before
----------------------------------------
Address API create failure when country_id parameter is abbreviation instead of numeric id

After
----------------------------------------
Address API create success when country_id parameter is abbreviation instead of numeric id
Fixes regression documented by Tim here: https://lab.civicrm.org/dev/core/issues/725#note_15384

Technical Details
----------------------------------------
See regression found by Tim before release: https://lab.civicrm.org/dev/core/issues/725#note_15384

Adding function _civicrm_api3_resolve_country_id() to api/v3/utils.php for the resolving of the country_id or country parameter to a numeric id for the country

Test provided by Tim included, we can close https://github.com/civicrm/civicrm-core/pull/13932/files after this passes.
Comments
----------------------------------------
Trying to make this work for the 'country' parameter, as well as supporting full country names as well as abbreviations if passed into the Address API create, so maybe more edge case regressions can be caught now, instead of reported later.

### CODE REVIEW

File: tests/phpunit/api/v3/AddressTest.php [hunk 1]
```diff
@@ -436,4 +436,43 @@ public function testCreateAddressStateProvinceIDCorrectForCountry() {
...
+    return [
+      // [mixed $inputCountry, mixed $inputState, int $expectCountry, int $expectState]
+      0 => [1228, 1004, 1228, 1004],
+      1 => ['US', 'CA', 1228, 1004],
+      2 => ['US', 'TX', 1228, 1042],
+      3 => ['US', 'California', 1228, 1004],
+      4 => [1228, 'Texas', 1228, 1042],
+      // Don't think these have been supported?
+      // 5 => ['United States', 1004, 1228, 1004] ,
+      // 6 => ['United States', 'TX', 1228, 1042],
```
Comment (colemanw) [inline tests/phpunit/api/v3/AddressTest.php]: One nitpick - there's really no reason to put the array keys 0, 1, 2, etc. in here as php will do it for you if you omit them.

Comment (jackrabbithanna) [reply tests/phpunit/api/v3/AddressTest.php]: added commit to remove the explicit numeric keys

**Comment by jackrabbithanna** [comment]
This test I got was provided by Tim. Is it even supported to pass state abbreviations in as the state_province_id? I think it's state name or id only? If this is the case we just update the test here and call it a day. I don't believe the test failures are related to the patches in question in this case then.

**Comment by eileenmcnaughton** [comment]
@jackrabbithanna & the test fails!

File: api/v3/utils.php [hunk 2]
```diff
@@ -2100,6 +2103,50 @@ function _civicrm_api3_validate_integer(&$params, $fieldName, &$fieldInfo, $enti
...
 }
 
+/**
+ * Helper function to determine country_id given the myriad of values for country_id or country that are supported
+ * @param $params
+ *
+ * @return int|null
+ */
+function _civicrm_api3_resolve_country_id($params) {
+  if (!empty($params['country_id'])) {
```
Comment (eileenmcnaughton) [inline api/v3/utils.php]: I think we have some cached functions that might be better BUT I'm happy to merge this when it passes tests & iterate for performance

File: tests/phpunit/api/v3/AddressTest.php (hunk 1 above)
Comment (totten) [reply tests/phpunit/api/v3/AddressTest.php]: There is a reason - when reporting failures, PHPUnit identifies the test case by row number. With unnumbered arrays, I've often found it annoying to determine which row it's talking about. With the numbers, it's easy.

**Comment by jackrabbithanna** [comment]
The test failures are the test that is being added by this PR. I'm not sure setting the state_province_id parameter with the state abbreviation is even working before these PRs....So I'm thinking I just update the test provided by this PR to not include data items with state_province_id values of abbreviation...

**Comment by totten** [comment]
Apologies - you're right, the abbreviations are not supported. (If they were, then `1 => [1228, 'CA', 1228, 1004],` would pass with or without fixes.) We can comment those out.

**Comment by jackrabbithanna** [comment]
I'm going to assume that test failure is unrelated ... REST API call getting a 503 response?

**Comment by seamuslee001** [comment]
@jackrabbithanna yes that would be right. Are you able to squish your commits into one then we can get this merged

**Comment by jackrabbithanna** [comment]
Commits squashed

**Comment by eileenmcnaughton** [comment]
thanks @jackrabbithanna

File: api/v3/utils.php (hunk 2 above)
Comment (jackrabbithanna) [reply api/v3/utils.php]: Oooh, please teach me the best way to cache a variable for the page request only in CiviCRM :) Do you'll use static variables like drupal_static() in D7 or got something else?

**Comment by colemanw** [comment]
@civicrm-builder test this please

File: tests/phpunit/api/v3/AddressTest.php (hunk 1 above)
Comment (colemanw) [reply tests/phpunit/api/v3/AddressTest.php]: Ahhh

**Comment by seamuslee001** [comment]
Merging as per the tag
